src/QsCompiler/CompilerOptimizations/Utils.fs

```diff
@@ -238,3 +130,82 @@ let rec internal takeWhilePlus1 (f: 'A -> bool) (l : list<'A>) =
+/// This function recurses through all the subscopes of the current scope.
+let rec private findAllBaseStatements (scope: QsScope): seq<QsStatementKind> =
+    scope.Statements |> Seq.collect (fun s ->
+        match s.Statement with
+        | QsConditionalStatement cond ->
+            seq {
+                for _, b in cond.ConditionalBlocks do yield b.Body
+                match cond.Default with
+                | Value b -> yield b.Body
+                | _ -> ()
+            } |> Seq.collect findAllBaseStatements
+        | QsForStatement {Body = scope}
+        | QsWhileStatement {Body = scope}
+        | QsQubitScope {Body = scope}
+        | QsScopeStatement {Body = scope} -> findAllBaseStatements scope
+        | QsRepeatStatement {RepeatBlock = scope1; FixupBlock = scope2} ->
+            Seq.concat [findAllBaseStatements scope1.Body; findAllBaseStatements scope2.Body]
+        | x -> Seq.singleton x
+    )
+
+/// Returns whether this scope contains any return statement
+let rec internal hasReturnStatement (scope: QsScope): bool =
+    scope |> findAllBaseStatements |> Seq.exists (function QsReturnStatement _ -> true | _ -> false)
+
+/// Returns the number of "bottom-level" statements in this scope
+let rec internal scopeLength (scope: QsScope): int =
+    scope |> findAllBaseStatements |> Seq.length
+
+
+/// Tries to inline the given expression, checking whether it's valid to do so.
+/// Returns the name of the callable and the modified scope after inlining.
+let internal tryInline (state: TransformationState) (ex: TypedExpression) =
+    maybe {
+        let! expr, arg =
+            match ex.Expression with
+            | CallLikeExpression ({Expression = expr}, arg) -> Some (expr, arg)
+            | _ -> None
+        let! qualName, specKind =
+            match expr with
+            | Identifier (GlobalCallable qualName, _) -> Some (qualName, QsBody)
+            | AdjointApplication {Expression = Identifier (GlobalCallable qualName, _)} -> Some (qualName, QsAdjoint)
+            | ControlledApplication {Expression = Identifier (GlobalCallable qualName, _)} -> Some (qualName, QsControlled)
+            | AdjointApplication {Expression = ControlledApplication {Expression = Identifier (GlobalCallable qualName, _)}} -> Some (qualName, QsControlledAdjoint)
+            | ControlledApplication {Expression = AdjointApplication {Expression = Identifier (GlobalCallable qualName, _)}} -> Some (qualName, QsControlledAdjoint)
+            | _ -> None
+        let callable = getCallable state qualName
+        let! impl = callable.Specializations |> Seq.tryFind (fun s -> s.Kind = specKind)
+        let! specArgs, scope =
+            match impl with
+            | {Implementation = Provided (specArgs, scope)} -> Some (specArgs, scope)
+            | {Implementation = Generated SelfInverse} ->
+                let newKind = match specKind with QsAdjoint -> Some QsBody | QsControlledAdjoint -> Some QsControlled | _ -> None
```

Comment:
Why do you only match two of the four cases?

Reply:
Because these are the only cases for which I can find the implementation - either it's Provided, or it's self-inverse and the inverse is Provided

Reply:
Sorry, I meant why do you return None if spedKind is not QsAdjoint or QsControlledAdjoint?